### PR TITLE
Add Namespace awareness for all available UDFs [HZ-2576]

### PR DESCRIPTION
**Pull request by JamesHazelcast**
This PR focuses on the changes required for the introduction of Namespace awareness to User-Defined Functions, primarily in relation to data structures, although "default" Namespace awareness is provided to UDFs not associated with a data structure.
[Namespaces feature TDD for reference](https://gist.github.com/JamesHazelcast/f25007a4e0594c491ce867c816b616a7).

_This PR is 1 of 5 pertaining to the Namespace feature addition. The other 4 PRs can be found below:_
- https://github.com/hazelcast/hazelcast/pull/26144
- https://github.com/hazelcast/hazelcast/pull/26145
- https://github.com/hazelcast/hazelcast/pull/26146
- https://github.com/hazelcast/hazelcast/pull/26147

_Due to the Namespaces feature being split into 5 pieces for easier reviewing, it is highly recommended that you clone the [UCD feature development branch](https://github.com/vbekiaris/hazelcast/tree/enhancements/5.4.0/ucd) for testing purposes, as these individual PRs are likely non-functional standalone._

### CODE REVIEW

**Review by frant-hartm — CHANGES_REQUESTED** [review]
Concurrent replacement of config - should the currently executing user code successfully finish?
It seems that the following can happen:
Run entry processor:
- processor is deserialized using namespace version 1
- replace namespace with version 2
- processor is executed using runWithNamespace version 2

IMap#replaceAll is not implemented with namespaces.

**Comment by JamesHazelcast** [comment]
> Concurrent replacement of config - should the currently executing user code successfully finish? It seems that the following can happen: Run entry processor:
> 
> * processor is deserialized using namespace version 1
> * replace namespace with version 2
> * processor is executed using runWithNamespace version 2

Yes this is the intended flow - we do not make any attempts to stop/interrupt/restart use of a Namespace when it is replaced, we rely on the next call using the updated Namespace instead.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 1]
```diff
@@ -529,6 +530,17 @@ public Object toObject(Object data) {
         }
     }
 
+    public Object toObject(Object data, String namespace) {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java:533]: There was already an existing `toObject` method that most/all cache operations were already using - so I just created a Namespace specific variant next to it to reduce duplicate code in all uses.

File: hazelcast/src/main/java/com/hazelcast/spi/annotation/NamespacesSupported.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Marker for interfaces that can be implemented by users and executed on clusters via User Code Deployment.
+ * Provides no functionality beyond being a marker for internal-use. Namely, this marker indicates that
+ * the class can safely be deserialized in member-to-member communication, with Namespace awareness.
+ *
+ * <b>Implementation note:</b> This marker only indicates <b>AP</b> data structure support.
+ */
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/spi/annotation/NamespacesSupported.java:32]: Good suggestion, thanks. Addressed in https://github.com/vbekiaris/hazelcast/commit/9001630d3b19a18d65bff86bfd811931b87a5dcb

**Review by vbekiaris — COMMENTED** [review]
I see no `getNamespacePermission` to `USE` namespace in Executor/ScheduledExecutor/DurableExecutor message tasks that submit a runnable/callable.
On EE side, we would need permissions tests - if we can take advantage of existing `UCDTest` children and test them with/without required client permissions, we would have caught those missing `getNamespacePermission()` overrides.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractMessageTask.java [hunk 3]
```diff
@@ -334,14 +336,18 @@ public String getDistributedObjectType() {
         return getServiceName();
     }
 
-    @Override
-    public abstract String getDistributedObjectName();
+    protected static Collection<Permission> extendPermissions(@Nullable Collection<Permission> parentPermissions,
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractMessageTask.java:339]: Yep! Thanks, addressed in https://github.com/vbekiaris/hazelcast/commit/b5185b6b8213d3cd2560670d07ada944804720c5

**Review by vbekiaris — APPROVED** [review]
✅ 

kudos @JamesHazelcast , @JackPGreen , @gareth-johnston

**Comment by JamesHazelcast** [comment]
Closing this PR due to the creation of the final mono-repo PR: https://github.com/hazelcast/hazelcast-mono/pull/217 🎉

**Pull request closed without merging**